crates/redb-bench/benches/common.rs

```diff
@@ -1408,3 +1409,182 @@ impl BenchInserter for FjallBenchInserter<'_, '_> {
+        SqliteBenchWriteTransaction {
+            conn: Arc::clone(&self.conn),
+        }
+    }
+
+    fn read_transaction(&self) -> Self::R<'_> {
+        SqliteBenchReadTransaction {
+            conn: Arc::clone(&self.conn),
+        }
+    }
+
+    fn compact(&mut self) -> bool {
+        let conn = self.conn.lock().unwrap();
+        conn.execute("VACUUM", []).unwrap();
+        true
+    }
+}
+
+pub struct SqliteBenchWriteTransaction {
+    conn: Arc<Mutex<Connection>>,
+}
+
+impl BenchWriteTransaction for SqliteBenchWriteTransaction {
+    type W<'txn>
+        = SqliteBenchInserter
+    where
+        Self: 'txn;
+
+    fn get_inserter(&mut self) -> Self::W<'_> {
+        SqliteBenchInserter {
+            conn: Arc::clone(&self.conn),
+        }
+    }
+
+    fn commit(self) -> Result<(), ()> {
+        Ok(())
+    }
+}
+
+pub struct SqliteBenchInserter {
+    conn: Arc<Mutex<Connection>>,
+}
+
+impl BenchInserter for SqliteBenchInserter {
+    fn insert(&mut self, key: &[u8], value: &[u8]) -> Result<(), ()> {
+        let conn = self.conn.lock().unwrap();
+        conn.execute(
+            "INSERT OR REPLACE INTO kv (key, value) VALUES (?, ?)",
+            [key, value],
+        )
+        .map(|_| ())
+        .map_err(|_| ())
+    }
+
+    fn remove(&mut self, key: &[u8]) -> Result<(), ()> {
+        let conn = self.conn.lock().unwrap();
+        conn.execute("DELETE FROM kv WHERE key = ?", [key])
+            .map(|_| ())
+            .map_err(|_| ())
+    }
+}
+
+pub struct SqliteBenchReadTransaction {
+    conn: Arc<Mutex<Connection>>,
+}
+
+impl BenchReadTransaction for SqliteBenchReadTransaction {
+    type T<'txn>
+        = SqliteBenchReader
+    where
+        Self: 'txn;
+
+    fn get_reader(&self) -> Self::T<'_> {
+        SqliteBenchReader {
+            conn: Arc::clone(&self.conn),
+        }
+    }
+}
+
+pub struct SqliteBenchReader {
+    conn: Arc<Mutex<Connection>>,
+}
+
+impl BenchReader for SqliteBenchReader {
+    type Output<'out>
+        = Vec<u8>
+    where
+        Self: 'out;
+    type Iterator<'out>
+        = SqliteBenchIterator
+    where
+        Self: 'out;
+
+    fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
+        let conn = self.conn.lock().unwrap();
+        let mut stmt = conn.prepare("SELECT value FROM kv WHERE key = ?").unwrap();
+        stmt.query_row([key], |row| row.get(0)).ok()
+    }
+
+    fn range_from<'a>(&'a self, key: &'a [u8]) -> Self::Iterator<'a> {
+        let conn = self.conn.lock().unwrap();
+        let mut stmt = conn
+            .prepare("SELECT key, value FROM kv WHERE key >= ? ORDER BY key")
+            .unwrap();
+        let rows = stmt
+            .query_map([key], |row| {
+                Ok((row.get::<_, Vec<u8>>(0)?, row.get::<_, Vec<u8>>(1)?))
+            })
+            .unwrap();
+
+        let mut results = Vec::new();
+        for row in rows {
+            if let Ok(kv) = row {
+                results.push(kv);
+            }
+        }
```

Comment:
Copying all the rows into the `results` vec is going to be very slow. I think the benchmark just reads a few, and then discards the rest. Is there a way to return `rows` as the iterator and fetch from the database on-demand?